src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+        const std::string& proof,
+        const uint256& pubKeyHash,
+        const uint256& randomSeed,
+        const boost::array<uint256, NumInputs>& macs,
+        const boost::array<uint256, NumInputs>& nullifiers,
+        const boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!vk) {
+            throw std::runtime_error("JoinSplit verifying key not loaded");
+        }
+
+        r1cs_ppzksnark_proof<ppzksnark_ppT> r1cs_proof;
+        std::stringstream ss;
+        ss.str(proof);
+        ss >> r1cs_proof;
+
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
+
+        auto witness = joinsplit_gadget<FieldT, NumInputs, NumOutputs>::witness_map(
+            rt,
+            h_sig,
+            macs,
+            nullifiers,
+            commitments,
+            vpub_old,
+            vpub_new
+        );
+
+        return r1cs_ppzksnark_verifier_strong_IC<ppzksnark_ppT>(*vk, witness, r1cs_proof);
+    }
+
+    std::string prove(
+        const boost::array<JSInput, NumInputs>& inputs,
+        const boost::array<JSOutput, NumOutputs>& outputs,
+        boost::array<Note, NumOutputs>& output_notes,
+        boost::array<ZCNoteEncryption::Ciphertext, NumOutputs>& ciphertexts,
+        uint256& ephemeralKey,
+        const uint256& pubKeyHash,
+        uint256& randomSeed,
+        boost::array<uint256, NumInputs>& macs,
+        boost::array<uint256, NumInputs>& nullifiers,
+        boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!pk) {
+            throw std::runtime_error("JoinSplit proving key not loaded");
+        }
+
+        // Compute nullifiers of inputs
+        for (size_t i = 0; i < NumInputs; i++) {
+            nullifiers[i] = inputs[i].nullifier();
+        }
+
+        // Sample randomSeed
+        randomSeed = random_uint256();
+
+        // Compute h_sig
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
+
+        // Sample phi
+        uint256 phi = random_uint256();
+
+        // Compute notes for outputs
+        for (size_t i = 0; i < NumOutputs; i++) {
+            // Sample r
+            uint256 r = random_uint256();
+
+            output_notes[i] = outputs[i].note(phi, r, i, h_sig);
+        }
+
+        // Compute the output commitments
+        for (size_t i = 0; i < NumOutputs; i++) {
+            commitments[i] = output_notes[i].cm();
+        }
+
+        // Encrypt the ciphertexts containing the note
+        // plaintexts to the recipients of the value.
+        {
+            ZCNoteEncryption encryptor(h_sig);
+
+            for (size_t i = 0; i < NumOutputs; i++) {
+                // TODO: expose memo in the public interface
+                boost::array<unsigned char, ZC_MEMO_SIZE> memo;
+                memo[0] = 0xF6; // invalid UTF8 as per spec
```

Comment:
According to boost docs, "passing no initializer list means that the elements have an indetermined initial value". The rest of the bytes of `memo` may be filled with uninitialized memory, which could potentially contain secrets which would get sent to whoever you're sending the note. Let's do:

```
boost::array<unsigned char, ZC_MEMO_SIZE> memo = { { 0xF6 } };
```

...which guarantees the other elements are default-initialized.

http://www.boost.org/doc/libs/1_60_0/doc/html/array/rationale.html